### PR TITLE
Idempotent binding creation

### PR DESCRIPTION
**Pull request by fwiesel**
In case we already created a binding for a live-migration
but crashed during the ongoing process, neutron will already
hold a port binding for the host.
Instead of failing, we can simply take the existing port-binding
and continue.

Possibly, also other port binding creations can profit from the behaviour.

### CODE REVIEW

**Comment by fwiesel** [comment]
I understand, you want rather smaller changes to review, so I break the live-migration down into changes, which are hopefully more easily reviewable.

File: nova/network/neutronv2/api.py [hunk 1]
```diff
@@ -1314,6 +1314,18 @@ def bind_ports_to_host(self, context, instance, host,
...
                 bindings_by_port_id[port_id] = resp.json()['binding']
             else:
+                # In case we actually managed to create the binding,
+                # but something went wrong in between, we can recover
+                if resp.status_code == 409:
+                    error_type = resp.json().get('NeutronError', {}).get('type')
+                    if error_type == 'PortBindingAlreadyExists':
+                        resp = client.get(
+                            '/v2.0/ports/{}/bindings/{}'.format(
+                                port_id, host))
```
Comment (joker-at-work) [inline nova/network/neutronv2/api.py:1324]: Do we need `raise_exc=False` here, too?

File: nova/network/neutronv2/api.py [hunk 2]
```diff
@@ -1314,6 +1314,18 @@ def bind_ports_to_host(self, context, instance, host,
             if resp:
                 bindings_by_port_id[port_id] = resp.json()['binding']
             else:
+                # In case we actually managed to create the binding,
+                # but something went wrong in between, we can recover
+                if resp.status_code == 409:
+                    error_type = resp.json().get('NeutronError', {}).get('type')
+                    if error_type == 'PortBindingAlreadyExists':
+                        resp = client.get(
+                            '/v2.0/ports/{}/bindings/{}'.format(
```
Comment (joker-at-work) [inline nova/network/neutronv2/api.py:1323]: I had a hard time finding that endpoint, as it's not in the networking documentation. I found it here https://specs.openstack.org/openstack/neutron-specs/specs/backlog/pike/portbinding_information_for_nova.html
I think we should put it into the commit-message for future reference.

File: nova/network/neutronv2/api.py [hunk 3]
```diff
@@ -1314,6 +1314,18 @@ def bind_ports_to_host(self, context, instance, host,
             if resp:
                 bindings_by_port_id[port_id] = resp.json()['binding']
             else:
+                # In case we actually managed to create the binding,
+                # but something went wrong in between, we can recover
+                if resp.status_code == 409:
+                    error_type = resp.json().get('NeutronError', {}).get('type')
+                    if error_type == 'PortBindingAlreadyExists':
+                        resp = client.get(
```
Comment (joker-at-work) [inline nova/network/neutronv2/api.py:1322]: With us overwriting `resp` here, we will get a confusing error-message, if the request fails for some reason. What's your opinion on doing an explicit `LOG.info`, that we will try to fetch the binding instead or something so we know were the `LOG.error` down below originated from?

**Comment by joker-at-work** [comment]
As I'd like to do a rollout next week, merging this PR has to wait, because the used endpoint is only available in Neutron > queens and the Neutron update is not done in all regions.

**Comment by fwiesel** [comment]
Sure, not problem. There is no hurry there. I still have to split up all the changes, which will take time.

**Comment by joker-at-work** [comment]
One more question: We yesterday discussed the topic of multiple bindings for a port and came to the conclusion, that pre-binding for a host is not possible with NSX-T < 3.2, which we don't have in production. Can you point me to the part where we hack around this to make live-migration work without packet-loss?

**Comment by fwiesel** [comment]
> Can you point me to the part where we hack around this to make live-migration work without packet-loss?

That doesn't exists. We cannot make it work without packet loss. With NSX-T <3.2, we do not know, which (NSX-T-)port the VM will have, until the VM is there, and only then will the NSX-T manager apply the rules. Until then, the default policy will apply, which is "block everything". 

What I've heard is, the upper limit for that to happen is two seconds, but we might want to check how it looks in production.

**Comment by joker-at-work** [comment]
> That doesn't exists.

Do we plan on implementing Csaba's hack to mitigate this? Or did he never find a working hack for the policy API? IIRC, the plan was to add a custom API to the nsxv3 driver (like we have for aci and asr1k already), that could trigger the hack-behavior.

**Comment by joker-at-work** [comment]
Do we know if we just drop packages and thus connections other than UDP can recover or do we reply with anything that would also close TCP connections?

**Review by joker-at-work — APPROVED** [review]

**Comment by fwiesel** [comment]
To my knowledge, the packages are simply dropped. So, TCP should recover. 
I'll check with Csaba on the other question.

**Comment by fwiesel** [comment]
So, the script had some port-binding hacks mostly for older versions of NSX-T (<3.0), where the port-binding was much more unreliable and with larger traffic interruptions. The way I understand it, we can only expect minor improvements without NSX-T 3.2 in the network connectivity.

I'll verify the tcp connectivity in a test.

**Pull request merged**